TowerScheduler/scheduler.py

```diff
@@ -0,0 +1,442 @@
+#!/usr/bin/python3
+'''
+This module contains our scheduler's state machine and associated states.
+Running the machine results in repeatedly executing the current state's process
+function, then executing its update function to proceed to the next state. The
+loop is ultimately broken when a call is made to a sleep timer's sleep function,
+at which point the sleep timer shuts down the machine running the state machine.
+'''
+
+from __future__ import annotations
+
+import argparse
+import datetime as dt
+import json
+import logging
+import sys
+import time
+from enum import Enum
+from pathlib import Path
+
+from ensemble import Ensemble
+from TowerScheduler.config import Configuration
+
+
+class CheckTimePath(Enum):
+    '''
+    Path to take from CheckTime state
+
+    skip: Time for current ensemble has been missed, so skip it (go to Iterate)
+    run: It's time to execute current ensemble (go to PerformEnsemble)
+    wait: Need to wait before current ensemble (go to Sleep)
+    reset: All ensembles for the day done, must reset (go to CheckTime for first
+            ensemble of next day)
+    '''
+    SKIP = 0
+    RUN = 1
+    WAIT = 2
+    RESET = 3
+
+def get_logger(name: str, level: int=logging.INFO):
+    logger = logging.getLogger(name)
+    logger.setLevel(logging.DEBUG)
+    handler = logging.StreamHandler(sys.stdout)
+    handler.setLevel(level)
+    formatter = logging.Formatter(
+            '%(levelname)s: %(name)s: %(message)s')
+    handler.setFormatter(formatter)
+    logger.addHandler(handler)
+    return logger
+
+
+class State:
+    '''
+    Although each state is divided into process and update, they aren't
+    always perfectly decoupled. Sometimes the update functioning has to
+    do a bit of processing to make its determination. The example of that
+    would be the Sleep state which calls into sleep then picks a new state.
+    '''
+
+    @classmethod
+    def get_singleton(cls) -> State:
+        '''
+        Return the singleton object for this state. If the singleton has not
+        been created yet, create one, set it as the singleton, and return it.
+        '''
+
+    def process(self, state_machine: StateMachine):
+        '''
+        The process function for each state does what the state is meant to
+        accomplish. To be overwritten in each state.
+
+        @param state_machine: the machine to which this state belongs
+        '''
+
+    def update(self, state_machine: StateMachine):
+        '''
+        The update function for each state selects the state to transition to.
+        To be overwritten in each state.
+
+        @param state_machine: the machine to which this state belongs
+        '''
```

Comment:
Would it make more sense from OOP to have update return the next state instead of modifying the state machine object?